File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.java
import java.util.*;

class Solution {
    
    static int N, M;
    static int dx[] = {0, -1, 0, 1}, dy[] = {1, 0, -1, 0};
    static boolean[][] visited;
    static int[][] mapInfo;
    
    public int[] solution(String[] maps) {
        
        N = maps.length;
        M = maps[0].length();

        mapInfo = new int[N][M];
        visited = new boolean[N][M];
        
        for(int i=0; i<N; i++){
            for(int j=0; j<M; j++){
                char input = maps[i].charAt(j);
                if(input == 'X'){
                    mapInfo[i][j] = 0;
                }else{
                    mapInfo[i][j] = input - '0';
                }
            }
        }
        
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        
        for(int i=0; i<N; i++){
            for(int j=0; j<M; j++){
                if(mapInfo[i][j] > 0 && !visited[i][j]){
                    System.out.println(i+ " " + j);
                    int sum = findIsland(i,j);
                    pq.add(sum);
                }
            }
        }
        
        if(pq.size() == 0) {
            return new int[]{-1};
        }
        
        int[] answer = new int[pq.size()];
        
        for(int i=0; i<answer.length; i++){
            answer[i] = pq.poll();
        }
        
        return answer;
    }
    
    public int findIsland(int x, int y){
        
        Deque<int[]> que = new ArrayDeque<>();
        que.add(new int[] {x, y});
        visited[x][y] = true;
        int sum = mapInfo[x][y];
        
        while(!que.isEmpty()){
            int now[] = que.poll();
            for(int i=0; i<4; i++){
                int nextX = now[0]+dx[i];
                int nextY = now[1]+dy[i];
                if(nextX < 0 || nextX >= N || nextY < 0 || nextY >= M) continue;
                if(visited[nextX][nextY]) continue;
                if(mapInfo[nextX][nextY] == 0) continue;
                sum += mapInfo[nextX][nextY];
                visited[nextX][nextY] = true;
                que.add(new int[]{nextX, nextY});
            }
        }
        
        return sum;
        
    }

}

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.js
var N, M, mapInfo, visited;
const dx = [0, -1, 0, 1];
const dy = [1, 0, -1, 0];


function solution(maps) {
    
    N = maps.length;
    M = maps[0].length;
    
    mapInfo = Array.from({length : N}, () => Array(M).fill(0));
    visited = Array.from({length : N}, () => Array(M).fill(false));
    
    for(let i=0; i<N; i++){
        for(let j=0; j<M; j++){
            if(maps[i].charAt(j) == 'X')continue;
            mapInfo[i][j] = maps[i].charAt(j) - '0';
        }
    }
        
    var answer = [];
    
    for(let i=0; i<N; i++){
        for(let j=0; j<M; j++){
            if(mapInfo[i][j] > 0 && !visited[i][j]){
                //console.log(i+" "+j)
                var sum = bfs(i , j);
                answer.push(sum);
            }
        }
    }
    
    if(answer.length == 0) return [-1]
    
    answer.sort((a,b) => a-b);
    
    return answer;
}

function bfs(x, y){
    
    const que = [];
    visited[x][y] = true;
    que.push([x, y]);
    var sum = mapInfo[x][y];
    
    while(que.length){
        var now = que.pop();
        for(let i=0; i<4; i++){
            var nextX = now[0] + dx[i];
            var nextY = now[1] + dy[i];
            if(nextX < 0 || nextX >= N || nextY < 0 || nextY >= M) continue;
            if(visited[nextX][nextY]) continue;
            if(mapInfo[nextX][nextY] == 0) continue;
            visited[nextX][nextY] = true;
            sum += mapInfo[nextX][nextY];
            que.push([nextX, nextY]);
        }
    }
    
    return sum;
    
}